File: Source/Client/Forms/FrmEditor_Animation.vb
Imports System.IO
Imports System.Windows.Forms
Imports Core
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Graphics

Friend Class frmEditor_Animation
    Private Sub NudSprite0_ValueChanged(sender As Object, e As EventArgs) Handles nudSprite0.Click
        Type.Animation(GameState.EditorIndex).Sprite(0) = nudSprite0.Value
    End Sub

    Private Sub NudSprite1_ValueChanged(sender As Object, e As EventArgs) Handles nudSprite1.Click
        Type.Animation(GameState.EditorIndex).Sprite(1) = nudSprite1.Value
    End Sub

    Private Sub NudLoopCount0_ValueChanged(sender As Object, e As EventArgs) Handles nudLoopCount0.Click
        Type.Animation(GameState.EditorIndex).LoopCount(0) = nudLoopCount0.Value
    End Sub

    Private Sub NudLoopCount1_ValueChanged(sender As Object, e As EventArgs) Handles nudLoopCount1.Click
        Type.Animation(GameState.EditorIndex).LoopCount(1) = nudLoopCount1.Value
    End Sub

    Private Sub NudFrameCount0_ValueChanged(sender As Object, e As EventArgs) Handles nudFrameCount0.Click
        Type.Animation(GameState.EditorIndex).Frames(0) = nudFrameCount0.Value
    End Sub

    Private Sub NudFrameCount1_ValueChanged(sender As Object, e As EventArgs) Handles nudFrameCount1.Click
        Type.Animation(GameState.EditorIndex).Frames(1) = nudFrameCount1.Value
    End Sub

    Private Sub NudLoopTime0_ValueChanged(sender As Object, e As EventArgs) Handles nudLoopTime0.Click
        Type.Animation(GameState.EditorIndex).LoopTime(0) = nudLoopTime0.Value
    End Sub

    Private Sub NudLoopTime1_ValueChanged(sender As Object, e As EventArgs) Handles nudLoopTime1.Click
        Type.Animation(GameState.EditorIndex).LoopTime(1) = nudLoopTime1.Value
    End Sub

    Private Sub BtnSave_Click(sender As Object, e As EventArgs) Handles btnSave.Click
        AnimationEditorOk()
        Dispose()
    End Sub

    Private Sub TxtName_TextChanged(sender As Object, e As EventArgs) Handles txtName.TextChanged
        Dim tmpindex As Integer
        tmpindex = lstIndex.SelectedIndex
        Type.Animation(GameState.EditorIndex).Name = Trim$(txtName.Text)
        lstIndex.Items.RemoveAt(GameState.EditorIndex - 1)
        lstIndex.Items.Insert(GameState.EditorIndex - 1, GameState.EditorIndex & ": " & Type.Animation(GameState.EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex
    End Sub

    Private Sub lstIndex_Click(sender As Object, e As MouseEventArgs)
        AnimationEditorInit()
    End Sub

    Private Sub BtnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
        Dim tmpindex As Integer

        ClearAnimation(GameState.EditorIndex)

        tmpindex = lstIndex.SelectedIndex
        lstIndex.Items.RemoveAt(GameState.EditorIndex - 1)
        lstIndex.Items.Insert(GameState.EditorIndex - 1, GameState.EditorIndex & ": " & Type.Animation(GameState.EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex

        AnimationEditorInit()
    End Sub

    Private Sub BtnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        AnimationEditorCancel()
        Dispose()
    End Sub

    Private Sub frmEditor_Animation_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        lstIndex.Items.Clear()

        ' Add the names
        For i = 1 To MAX_ANIMATIONS
            lstIndex.Items.Add(i & ": " & Type.Animation(i).Name)
        Next

        ' find the music we have set
        cmbSound.Items.Clear()

        CacheSound()

        For i = 1 To UBound(SoundCache)
            cmbSound.Items.Add(SoundCache(i))
        Next

        nudSprite0.Maximum = GameState.NumAnimations
        nudSprite1.Maximum = GameState.NumAnimations
    End Sub

    Private Sub CmbSound_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbSound.SelectedIndexChanged
        Type.Animation(GameState.EditorIndex).Sound = cmbSound.SelectedItem.ToString
    End Sub

    Private Sub frmEditor_Animation_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        AnimationEditorCancel()
    End Sub

    Public Sub ProcessAnimation(animationControl As System.Windows.Forms.NumericUpDown,
                            frameCountControl As System.Windows.Forms.NumericUpDown,
                            loopCountControl As System.Windows.Forms.NumericUpDown,
                            animationTimerIndex As Integer,
                            renderTarget As RenderTarget2D,
                            backgroundColorControl As System.Windows.Forms.PictureBox,
                            spriteBatch As Graphics.SpriteBatch)

        ' Retrieve the animation number and check its validity
        Dim animationNum As Integer = animationControl.Value
        If animationNum <= 0 Or animationNum > GameState.NumAnimations Then
            spriteBatch.GraphicsDevice.Clear(GameClient.ToMonoGameColor(backgroundColorControl.BackColor))
            Exit Sub
        End If

        ' Check whether animationDisplay is Texture2D or System.Drawing.Image
        Dim texture As Graphics.Texture2D
        texture = GameClient.GetTexture(System.IO.Path.Combine(Core.Path.Animations, animationNum & GameState.GfxExt))
        If texture Is Nothing Then
            Exit Sub
        End If

        ' Get dimensions and column count from controls and graphic info
        Dim totalWidth As Integer
        Dim totalHeight As Integer

        Dim gfxInfo = GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Animations, animationNum & GameState.GfxExt))
        totalWidth = gfxInfo.Width
        totalHeight = gfxInfo.Height

        Dim columns As Integer = frameCountControl.Value

        ' Validate columns to avoid division by zero
        If columns <= 0 Then Exit Sub

        ' Calculate frame dimensions
        Dim frameWidth As Integer = totalWidth / columns

        ' Assuming square frames for simplicity (adjust if frames are not square)
        Dim frameHeight As Integer = frameWidth

        Dim rows As Integer
        If frameHeight > 0 Then
            rows = totalHeight / frameHeight
        End If

        Dim frameCount As Integer = rows * columns

        ' Retrieve loop timing and check frame rendering necessity
        Dim looptime As Integer = loopCountControl.Value
        If GameState.AnimEditorTimer(animationTimerIndex) + looptime <= Environment.TickCount Then
            If GameState.AnimEditorFrame(animationTimerIndex) >= frameCount Then
                GameState.AnimEditorFrame(animationTimerIndex) = 1 ' Reset to the first frame if it exceeds the count
            Else
                GameState.AnimEditorFrame(animationTimerIndex) += 1
            End If
            GameState.AnimEditorTimer(animationTimerIndex) = Environment.TickCount

            ' Render the frame if necessary
            If frameCountControl.Value > 0 Then
                Dim frameIndex As Integer = GameState.AnimEditorFrame(animationTimerIndex) - 1
                Dim column As Integer = frameIndex Mod columns
                Dim row As Integer = frameIndex \ columns

                ' Calculate the source rectangle for the texture or image
                Dim sRECT As New Rectangle(column * frameWidth, row * frameHeight, frameWidth, frameHeight)

                ' Clear the background to the specified color
                spriteBatch.GraphicsDevice.Clear(GameClient.ToMonoGameColor(backgroundColorControl.BackColor))

                GameClient.Graphics.GraphicsDevice.SetRenderTarget(renderTarget)

                ' Draw MonoGame texture
                spriteBatch.Begin()
                spriteBatch.Draw(texture, New Rectangle(0, 0, frameWidth, frameHeight), sRECT, Color.White)
                spriteBatch.End()

                GameClient.Graphics.GraphicsDevice.SetRenderTarget(Nothing)

                ' Convert the render target to a Texture2D and set it as the PictureBox background
                Using stream As New System.IO.MemoryStream()
                    renderTarget.SaveAsPng(stream, renderTarget.Width, renderTarget.Height)
                    stream.Position = 0
                    backgroundColorControl.Image = Drawing.Image.FromStream(stream)
                End Using
            End If
        End If
    End Sub

    Private Sub picSprite0_Paint(sender As Object, e As PaintEventArgs) Handles picSprite0.Paint
        DrawAnimationSprite0()
    End Sub

    Private Sub picSprite1_Paint(sender As Object, e As PaintEventArgs) Handles picSprite1.Paint
        DrawAnimationSprite1()
    End Sub

    Private Sub DrawAnimationSprite0()
        With frmEditor_Animation.Instance
            ' Ensure spriteBatch is created and disposed properly
            Dim spriteBatch As New Graphics.SpriteBatch(GameClient.Graphics.GraphicsDevice)
            Dim renderTarget As New RenderTarget2D(GameClient.Graphics.GraphicsDevice, .picSprite0.Width, .picSprite0.Height)

            ' Call ProcessAnimation for each animation panel
            ProcessAnimation(.nudSprite0, .nudFrameCount0, .nudLoopTime0, 0, renderTarget, .picSprite0, spriteBatch)
        End With
    End Sub

    Private Sub DrawAnimationSprite1()
        With frmEditor_Animation.Instance
            ' Ensure spriteBatch is created and disposed properly
            Dim spriteBatch As New Graphics.SpriteBatch(GameClient.Graphics.GraphicsDevice)
            Dim renderTarget As New RenderTarget2D(GameClient.Graphics.GraphicsDevice, .picSprite1.Width, .picSprite1.Height)

            ' Call ProcessAnimation for each animation panel
            ProcessAnimation(.nudSprite1, .nudFrameCount1, .nudLoopTime1, 1, renderTarget, .picSprite1, spriteBatch)
        End With
    End Sub
End Class

File: Source/Client/Forms/FrmEditor_Animation.cs
using Core;
using DarkUI.Controls;
using Microsoft.VisualBasic;
using Microsoft.Xna.Framework.Graphics;
using System.Linq.Expressions;
using System.Reflection.Metadata;
using Color = Microsoft.Xna.Framework.Color;
using Rectangle = Microsoft.Xna.Framework.Rectangle;
using RenderTargetUsage = Microsoft.Xna.Framework.Graphics.RenderTargetUsage;

namespace Client
{
    internal partial class frmEditor_Animation
    {

        protected override void WndProc(ref Message m)
        {
            const int WM_MOUSEACTIVATE = 0x0021;
            const int WM_NCHITTEST = 0x0084;

            if (m.Msg == WM_MOUSEACTIVATE)
            {
                // Immediately activate and process the click.
                m.Result = new IntPtr(1); // MA_ACTIVATE
                return;
            }
            else if (m.Msg == WM_NCHITTEST)
            {
                // Let the window know the mouse is in client area.
                m.Result = new IntPtr(1); // HTCLIENT
                return;
            }

            base.WndProc(ref m);
        }

        private void NudSprite0_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].Sprite[0] = (int)Math.Round(nudSprite0.Value);
        }

        private void NudSprite1_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].Sprite[1] = (int)Math.Round(nudSprite1.Value);
        }

        private void NudLoopCount0_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].LoopCount[0] = (int)Math.Round(nudLoopCount0.Value);
        }

        private void NudLoopCount1_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].LoopCount[1] = (int)Math.Round(nudLoopCount1.Value);
        }

        private void NudFrameCount0_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].Frames[0] = (int)Math.Round(nudFrameCount0.Value);
        }

        private void NudFrameCount1_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].Frames[1] = (int)Math.Round(nudFrameCount1.Value);
        }

        private void NudLoopTime0_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].LoopTime[0] = (int)Math.Round(nudLoopTime0.Value);
        }

        private void NudLoopTime1_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].LoopTime[1] = (int)Math.Round(nudLoopTime1.Value);
        }

        private void BtnSave_Click(object sender, EventArgs e)
        {
            Editors.AnimationEditorOK();
            Dispose();
        }

        private void TxtName_TextChanged(object sender, EventArgs e)
        {
            int tmpindex;
            tmpindex = lstIndex.SelectedIndex;
            Core.Type.Animation[GameState.EditorIndex].Name = Strings.Trim(txtName.Text);
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Type.Animation[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;
        }

        private void lstIndex_Click(object sender, EventArgs e)
        {
            Editors.AnimationEditorInit();
        }

        private void BtnDelete_Click(object sender, EventArgs e)
        {
            int tmpindex;

            Animation.ClearAnimation(GameState.EditorIndex);

            tmpindex = lstIndex.SelectedIndex;
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Type.Animation[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;

            Editors.AnimationEditorInit();
        }

        private void BtnCancel_Click(object sender, EventArgs e)
        {
            Editors.AnimationEditorCancel();
            Dispose();
        }

        private void frmEditor_Animation_Load(object sender, EventArgs e)
        {
            lstIndex.Items.Clear();

            // Add the names
            for (int i = 0; i < Core.Constant.MAX_ANIMATIONS; i++)
                lstIndex.Items.Add(i + 1 + ": " + Core.Type.Animation[i].Name);

            // find the music we have set
            cmbSound.Items.Clear();

            General.CacheSound();

            for (int i = 0, loopTo = Information.UBound(Sound.SoundCache); i < loopTo; i++)
                cmbSound.Items.Add(Sound.SoundCache[i]);

            nudSprite0.Maximum = GameState.NumAnimations;
            nudSprite1.Maximum = GameState.NumAnimations;
        }

        private void CmbSound_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Animation[GameState.EditorIndex].Sound = cmbSound.SelectedItem.ToString();
        }

        private void frmEditor_Animation_FormClosing(object sender, FormClosingEventArgs e)
        {
            Editors.AnimationEditorCancel();
        }

        public void ProcessAnimation(ref DarkNumericUpDown animationControl, ref DarkNumericUpDown frameCountControl, ref DarkNumericUpDown loopCountControl, int animationTimerIndex, RenderTarget2D renderTarget, ref PictureBox backgroundColorControl, Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch)
        {
            try
            {
                int animationNum = (int)Math.Round(animationControl.Value);
                if (animationNum <= 0 || animationNum > GameState.NumAnimations)
                {
                    if (backgroundColorControl.Image != null)
                        backgroundColorControl.Image = null;
                    return;
                }

                var imagePath = System.IO.Path.Combine(Core.Path.Animations, animationNum + GameState.GfxExt);
                if (!System.IO.File.Exists(imagePath))
                {
                    backgroundColorControl.Image = null;
                    return;
                }

                using (var img = System.Drawing.Image.FromFile(imagePath))
                {
                    int columns = (int)Math.Round(frameCountControl.Value);
                    if (columns <= 0)
                    {
                        backgroundColorControl.Image = (System.Drawing.Image)img.Clone();
                        return;
                    }

                    int frameWidth = img.Width / columns;
                    int frameHeight = img.Height;
                    int rows = frameHeight > 0 ? img.Height / frameHeight : 1;
                    int frameCount = rows * columns;

                    int looptime = (int)Math.Round(loopCountControl.Value);
                    if (GameState.AnimEditorTimer[animationTimerIndex] + looptime <= Environment.TickCount)
                    {
                        if (GameState.AnimEditorFrame[animationTimerIndex] >= frameCount)
                        {
                            GameState.AnimEditorFrame[animationTimerIndex] = 1;
                        }
                        else
                        {
                            GameState.AnimEditorFrame[animationTimerIndex] += 1;
                        }
                        GameState.AnimEditorTimer[animationTimerIndex] = Environment.TickCount;
                    }

                    if (frameCountControl.Value > 0m)
                    {
                        int frameIndex = GameState.AnimEditorFrame[animationTimerIndex] - 1;
                        int column = frameIndex % columns;
                        int row = frameIndex / columns;

                        var sRECT = new System.Drawing.Rectangle(column * frameWidth, row * frameHeight, frameWidth, frameHeight);
                        var bmp = new System.Drawing.Bitmap(frameWidth, frameHeight);
                        using (var g = System.Drawing.Graphics.FromImage(bmp))
                        {
                            g.Clear(backgroundColorControl.BackColor);
                            g.DrawImage(img, new System.Drawing.Rectangle(0, 0, frameWidth, frameHeight), sRECT, System.Drawing.GraphicsUnit.Pixel);
                        }
                        backgroundColorControl.Image = bmp;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error processing animation: {ex.Message}");
                backgroundColorControl.Image = null;
            }
        }

        private void picSprite0_Paint(object sender, PaintEventArgs e)
        {
            // Call ProcessAnimation for Sprite0
            ProcessAnimation(
                ref nudSprite0,
                ref nudFrameCount0,
                ref nudLoopTime0,
                0,
                null,
                ref picSprite0,
                null
            );
        }

        private void picSprite1_Paint(object sender, PaintEventArgs e)
        {
            // Call ProcessAnimation for Sprite1
            ProcessAnimation(
                ref nudSprite1,
                ref nudFrameCount1,
                ref nudLoopTime1,
                1,
                null,
                ref picSprite1,
                null
            );
        }
    }
}